tour_model_eval/trips_in_bins_and_clusters.ipynb

```diff
@@ -0,0 +1,344 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "sonic-essay",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "########clusters#####"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "corresponding-silly",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "feat = featurization.featurization(bin_trips)\n",
+    "feat"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "lightweight-plaintiff",
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "# cluster the data using k-means\n",
+    "# def cluster(data, bins)\n",
+    "# return feat.clusters, feat.labels, feat.data\n",
+    "clusters,labels,feat_data = pipeline.cluster(bin_trips,len(bins))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "perfect-lighter",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Copy from featurization.py, should later change it to use folium\n",
+    "def map_clusters(self):\n",
+    "#         import matplotlib.pyplot as plt\n",
+    "#         from matplotlib import colors as matcol\n",
+    "#         colormap = plt.cm.get_cmap()\n",
+    "\n",
+    "    cluster_map = folium.Map(location=[self.points[0][1], self.points[0][0]],zoom_start=12, max_zoom= 30, control_scale=True)\n",
+    "\n",
+    "\n",
+    "    if self.labels:             \n",
+    "        for i in range(len(self.points)):\n",
+    "            start_lat = self.points[i][1]\n",
+    "            start_lon = self.points[i][0]\n",
+    "            end_lat = self.points[i][3]\n",
+    "            end_lon = self.points[i][2]\n",
+    "            path = [(start_lat, start_lon), (end_lat, end_lon)]\n",
+    "            layer = folium.PolyLine([[start_lat,start_lon],\n",
+    "                                     [end_lat,end_lon]],weight=2, color='orange')\n",
+    "            layer.add_to(cluster_map)\n",
+    "    cluster_map"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "reflected-second",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# clusters"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "varying-european",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# labels"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "blocked-biotechnology",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "# feat_data"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "durable-devon",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# feat.points"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "occasional-basis",
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "# This doesn't show the map, I don't know why.\n",
+    "map_clusters(feat)"
```

Comment:
because you weren't returning the cluster map.
In the previous viz, you didn't have a function to display the map, you did it directly in a cell
so typing out `m` at the end was enough

But here, you are calling a function. The function needs to return the map otherwise the calling cell doesn't get it.
Pretty straightforward.

Reply:
Oh I see!!!
But I still don't know how to plot the map based on labels? `clusters`after running `pipeline.cluster`is just the number of cluster. But the points are in a list with their own orders. How can they connect to the labels?

Reply:
Correct, this is a necessary but not a sufficient fix. Please see my next comment for the level at which you need to work with the code.